Sources/DistributedSystem/ChannelCompression.swift

```diff
@@ -0,0 +1,202 @@
+            }
+            _ = context.pipeline.removeHandler(self)
+        } else {
+            let channel = context.channel
+            logger.info("Invalid compression request received from client @ \(channel.remoteAddressDescription), close connection.")
+            channel.close(promise: nil)
+        }
+    }
+}
+
+final class ChannelCompressionHandshakeClient: ChannelInboundHandler, RemovableChannelHandler {
+    typealias InboundIn = ByteBuffer
+
+    private let loggerBox: Box<Logger>
+    private let compression: UInt8
+    private let streamHandler: NIOCore.ChannelHandler
+
+    init(_ loggerBox: Box<Logger>, _ compression: UInt8, _ streamHandler: NIOCore.ChannelHandler) {
+        self.loggerBox = loggerBox
+        self.compression = compression
+        self.streamHandler = streamHandler
+    }
+
+    func channelActive(context: ChannelHandlerContext) {
+        var buffer = ByteBufferAllocator().buffer(capacity: MemoryLayout<UInt8>.size)
+        buffer.writeInteger(compression, as: UInt8.self)
+        context.writeAndFlush(NIOAny(buffer), promise: nil)
+        if compression != 0 {
+            _ = context.pipeline.addHandler(ChannelCompressionInboundHandler(loggerBox), position: .after(streamHandler))
+            _ = context.pipeline.addHandler(ChannelCompressionOutboundHandler(), position: .last)
+        }
+        context.fireChannelActive()
+        _ = context.pipeline.removeHandler(self)
+    }
+}
+
+fileprivate struct BufferManager {
+    var buffers = Array(repeating: UnsafeMutableRawBufferPointer(start: nil, count: 0), count: 2)
+    var nextIdx = 0
+
+    mutating func getNextBuffer(capacity: Int) -> UnsafeMutableRawBufferPointer {
+        let idx = nextIdx
+        nextIdx = (1 - nextIdx)
+        var buffer = buffers[idx]
+        if buffer.count < capacity {
+            buffer.deallocate()
+            buffer = UnsafeMutableRawBufferPointer.allocate(byteCount: capacity, alignment: 0)
+            buffers[idx] = buffer
+        }
+        return buffer
+    }
+
+    mutating func reset() {
+        for idx in 0..<buffers.count {
+            buffers[idx].deallocate()
+        }
+    }
+}
+
+final class ChannelCompressionInboundHandler: ChannelInboundHandler {
+    typealias InboundIn = ByteBuffer
+    typealias InboundOut = ByteBuffer
+
+    private let loggerBox: Box<Logger>
+    private var bufferManager = BufferManager()
+    private var lz4Stream = LZ4_streamDecode_t()
+
+    private var logger: Logger { loggerBox.value }
+
+    init(_ loggerBox: Box<Logger>) {
+        self.loggerBox = loggerBox
+        LZ4_setStreamDecode(&lz4Stream, nil, 0);
+    }
+
+    deinit {
+        bufferManager.reset()
+    }
+
+    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        var bufferIn = unwrapInboundIn(data)
+        _ = bufferIn.readInteger(as: UInt32.self) // skip message size
+        if let messageSize = bufferIn.readInteger(as: UInt32.self) {
```

Comment:
I thought about it, but we should set message size in the beginning of the message and without knowing the amount of compressed data we can't reserve the space properly for that. Probably we could reserve maximum possible space, but code complexity in that case will grow. I decided to sacrifice the size of the data for the sake of code simplicity. We can fix it later if it will be a problem.

Reply:
But you anyway compress to auxiliary buffer due to LZ4 requirements and then copy. [](edited-with-Axolo)